File: distributed/kakiage/tensor_serializer.py
from typing import Dict, List, Tuple, Union
import struct
from dataclasses import dataclass
import numpy as np
from .constant_codec_eightbit import compress_tensor_eightbit

FILE_SIGNATURE = b"WDN2"
TENSOR_SIGNATURE = b"TENS"
CLOSE_SIGNATURE = b"CLOS"

DATA_TYPE_TO_NUMPY = {
    1: np.float32,  # onnx.TensorProto.FLOAT
    2: np.uint8,  # onnx.TensorProto.UINT8
    3: np.int8,  # onnx.TensorProto.INT8
    4: np.uint16,  # onnx.TensorProto.UINT16
    5: np.int16,  # onnx.TensorProto.INT16
    6: np.int32,  # onnx.TensorProto.INT32
    7: np.int64,  # onnx.TensorProto.INT64
    9: np.bool,  # onnx.TensorProto.BOOL
    10: np.float16,  # onnx.TensorProto.FLOAT16
    11: np.float64,  # onnx.TensorProto.DOUBLE
    12: np.uint32,  # onnx.TensorProto.UINT32
    13: np.uint64,  # onnx.TensorProto.UINT64
}


def _data_type_from_numpy(np_dtype) -> int:
    # dict like {np.float32: 1} cannot be used due to key equality check
    for k, v in DATA_TYPE_TO_NUMPY.items():
        if v == np_dtype:
            return k
    raise ValueError


def _compress_tensor_raw(data: np.ndarray) -> bytes:
    return data.tobytes()


def _compress_tensor(data: np.ndarray, compression_algorithm: int) -> bytes:
    if compression_algorithm == 0:
        return _compress_tensor_raw(data)
    elif compression_algorithm == 1:
        return compress_tensor_eightbit(data)
    else:
        raise ValueError


def _select_compression_algorithm(data: np.ndarray, compression_algorithm: int) -> int:
    if data.dtype != np.float32:
        return 0
    return compression_algorithm


def _make_tensor_chunk(name: str, data: np.ndarray, compression_algorithm: int) -> bytes:
    data_type = _data_type_from_numpy(data.dtype)
    compression_algorithm = _select_compression_algorithm(
        data, compression_algorithm)
    compressed_body = _compress_tensor(data, compression_algorithm)
    compressed_body_size = len(compressed_body)
    ndim = data.ndim
    dims = data.shape
    name_bytes = name.encode("utf-8")
    name_length = len(name_bytes)
    extra_bytes = b""
    extra_length = len(extra_bytes)
    header = struct.pack("<BIBB",
                         compression_algorithm,
                         compressed_body_size,
                         data_type,
                         ndim)
    header += struct.pack("<" + "I" * len(dims), *dims)
    header += struct.pack("<I", name_length)
    header += name_bytes
    header += struct.pack("<I", extra_length)
    header += extra_bytes
    header += compressed_body
    header = TENSOR_SIGNATURE + struct.pack("<I", len(header)) + header
    return header


def _make_close_chunk() -> bytes:
    return CLOSE_SIGNATURE + b"\0\0\0\0"


def serialize_tensors_to_bytes(tensors: Dict[str, np.ndarray], compression_algorithm: int = 0):
    chunks = [FILE_SIGNATURE]
    for name, data in tensors.items():
        chunks.append(_make_tensor_chunk(name, data, compression_algorithm))
    chunks.append(_make_close_chunk())
    full_data = b"".join(chunks)
    return full_data


def serialize_tensors(path_template: str, tensors: Dict[str, np.ndarray], split_size: int = 0, compression_algorithm: int = 0) -> List[str]:
    full_data = serialize_tensors_to_bytes(tensors, compression_algorithm)
    if split_size <= 0:
        with open(path_template, "wb") as f:
            f.write(full_data)
        return [path_template]
    else:
        file_paths = []
        for i in range((len(full_data) + split_size - 1) // split_size):
            file_path = path_template.format(i)
            file_paths.append(file_path)
            with open(file_path, "wb") as f:
                f.write(full_data[i*split_size:(i+1)*split_size])
        return file_paths


def deserialize_tensor(paths: Union[str, List[str]]) -> Dict[str, np.ndarray]:
    data_all = []
    if isinstance(paths, list):
        p = paths
    else:
        p = [paths]
    for path in p:
        with open(path, "rb") as f:
            data_all.append(f.read())
    return deserialize_tensor_from_bytes(b"".join(data_all))


@dataclass
class ChunkInfo:
    signature: bytes
    body: bytes
    next_offset: int


def deserialize_tensor_from_bytes(data: bytes) -> Dict[str, np.ndarray]:
    if data[:4] != FILE_SIGNATURE:
        raise ValueError("Unexpected file signature")
    offset = 4
    close = False
    tensors = {}
    while not close:
        chunk_info = _extract_chunk(data, offset)
        if chunk_info.signature == TENSOR_SIGNATURE:
            name, array = _parse_tensor_chunk(chunk_info.body)
            tensors[name] = array
        elif chunk_info.signature == CLOSE_SIGNATURE:
            close = True
        offset = chunk_info.next_offset
    return tensors


def _extract_chunk(data: bytes, offset: int) -> ChunkInfo:
    if len(data) < offset + 8:
        raise ValueError("Unexpected EOF")
    signature = data[offset:offset+4]
    body_byte_length = struct.unpack("<I", data[offset+4:offset+8])[0]
    if len(data) < offset + 8 + body_byte_length:
        raise ValueError("Unexpected EOF")
    return ChunkInfo(signature=signature, body=data[offset+8:offset+8+body_byte_length], next_offset=offset+8+body_byte_length)


def _parse_tensor_chunk(body: bytes) -> Tuple[str, np.ndarray]:
    offset = 0
    compression_algorithm, body_compressed_length, data_type, ndim = struct.unpack(
        "<BIBB", body[offset:offset+7])
    offset += 7
    dims = []
    for _ in range(ndim):
        dims.append(struct.unpack("<I", body[offset:offset+4])[0])
        offset += 4
    name_length = struct.unpack("<I", body[offset:offset+4])[0]
    offset += 4
    name = body[offset:offset+name_length].decode("utf-8")
    offset += name_length
    extra_length = struct.unpack("<I", body[offset:offset+4])[0]
    offset += 4
    # extra領域は未使用
    offset += extra_length

    if compression_algorithm != 0:
        raise ValueError("Compression algorithm not implemented")
    numpy_dtype = DATA_TYPE_TO_NUMPY[data_type]
    array = np.frombuffer(
        body[offset:offset+body_compressed_length], dtype=numpy_dtype).reshape(dims)
    return name, array

File: src/tensor/serializer/tensorSerializer.ts
import { DType, TypedArrayTypes } from '../../dtype';
import {
  arrayProd,
  arraySum,
  base64ToUint8Array,
  uint8ArrayToBase64,
} from '../../util';
import { CPUTensor } from '../cpu/cpuTensor';
import { decodeTensorRaw } from './tensorDecoder/decodeTensorRaw';

const signatureFile = 843990103, // "WDN2"
  signatureTensor = 1397638484, // "TENS"
  signautreClose = 1397705795; // "CLOS"

export enum TensorSerializerDataType {
  UNDEFINED = 0,
  FLOAT = 1,
  UINT8 = 2,
  INT8 = 3,
  UINT16 = 4,
  INT16 = 5,
  INT32 = 6,
  INT64 = 7,
  STRING = 8,
  BOOL = 9,
  FLOAT16 = 10,
  DOUBLE = 11,
  UINT32 = 12,
  UINT64 = 13,
  COMPLEX64 = 14,
  COMPLEX128 = 15,
  BFLOAT16 = 16,
}

const localStoragePrefix = 'localstorage://';

export class TensorDeserializer {
  async fromHTTP(paths: string[] | string): Promise<Map<string, CPUTensor>> {
    let pathsArray: string[];
    if (typeof paths === 'string') {
      pathsArray = [paths];
    } else {
      pathsArray = paths;
    }
    // TODO: progress
    const fileArray = await this.fetchAllFile(pathsArray);
    return this.deserialize(fileArray);
  }

  async fromLocalStorage(path: string): Promise<Map<string, CPUTensor>> {
    if (!path.startsWith(localStoragePrefix)) {
      throw new Error(
        `toLocalStorage: path must be prefix '${localStoragePrefix}'`
      );
    }
    const key = path.substring(localStoragePrefix.length) + '/0';
    const serialized = localStorage.getItem(key);
    if (serialized) {
      return this.deserialize(base64ToUint8Array(serialized));
    } else {
      throw new Error('no tensor stored');
    }
  }

  /**
   * Load from local file (e.g. files[0] of <input type="file">)
   * @param file
   * @returns
   */
  async fromFile(file: File | Blob): Promise<Map<string, CPUTensor>> {
    const ab = await file.arrayBuffer();
    return this.deserialize(new Uint8Array(ab));
  }

  deserialize(data: Uint8Array): Map<string, CPUTensor> {
    const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
    if (signatureFile !== view.getUint32(0, true)) {
      throw new Error('Unexpected file signature');
    }
    let offset = 4;
    const tensors = new Map<string, CPUTensor>();
    let close = false;
    while (!close) {
      const chunkInfo = this.extractChunk(data.buffer, offset);
      switch (chunkInfo.signature) {
        case signatureTensor:
          {
            const { name, tensor } = this.parseTensorChunk(
              data.buffer,
              chunkInfo.bodyByteOffset,
              chunkInfo.bodyByteLength
            );
            tensors.set(name, tensor);
          }
          break;
        case signautreClose:
          close = true;
          break;
      }
      offset = chunkInfo.nextByteOffset;
    }
    return tensors;
  }

  private async fetchAllFile(paths: string[]): Promise<Uint8Array> {
    // TODO: HTTP fetch以外への対応(localStorage等)
    const abs: ArrayBuffer[] = [];
    for (const path of paths) {
      const f = await fetch(path),
        ab = await f.arrayBuffer();
      abs.push(ab);
    }
    const totalLength = arraySum(abs.map((ab) => ab.byteLength)),
      concatArray = new Uint8Array(totalLength);
    let ofs = 0;
    for (const ab of abs) {
      const src = new Uint8Array(ab);
      concatArray.set(src, ofs);
      ofs += src.byteLength;
    }

    return concatArray;
  }

  private extractChunk(
    buf: ArrayBuffer,
    byteOffset: number
  ): {
    signature: number;
    nextByteOffset: number;
    bodyByteOffset: number;
    bodyByteLength: number;
  } {
    const view = new DataView(buf, byteOffset);
    if (view.byteLength < 8) {
      throw new Error('Unexpected EOF');
    }
    const signature = view.getUint32(0, true),
      bodyByteLength = view.getUint32(4, true),
      bodyByteOffset = byteOffset + 8;
    if (view.byteLength < 8 + bodyByteLength) {
      throw new Error('Unexpected EOF');
    }
    const nextByteOffset = bodyByteOffset + bodyByteLength;
    return { signature, bodyByteLength, bodyByteOffset, nextByteOffset };
  }

  private parseTensorChunk(
    buf: ArrayBuffer,
    bodyByteOffset: number,
    bodyByteLength: number
  ): { name: string; tensor: CPUTensor } {
    const view = new DataView(buf, bodyByteOffset, bodyByteLength);

    let ofs = 0;
    const compressionAlgorithm = view.getUint8(ofs);
    ofs += 1;
    const bodyCompressedLength = view.getUint32(ofs, true);
    ofs += 4;
    const dataType = view.getUint8(ofs);
    ofs += 1;
    const ndim = view.getUint8(ofs);
    ofs += 1;
    const dims: number[] = [];
    for (let i = 0; i < ndim; i++) {
      dims.push(view.getUint32(ofs, true));
      ofs += 4;
    }
    const numel = arrayProd(dims),
      nameLength = view.getUint32(ofs, true);
    ofs += 4;
    const name = this.parseString(buf, bodyByteOffset + ofs, nameLength);
    ofs += nameLength;
    const extraLength = view.getUint32(ofs, true);
    ofs += 4;
    // Skip extra data
    ofs += extraLength;

    const data = this.parseTensorBody(
      buf,
      compressionAlgorithm,
      bodyByteOffset + ofs,
      bodyCompressedLength,
      dataType,
      numel
    );
    let dataTypeString: DType;
    switch (dataType) {
      case TensorSerializerDataType.FLOAT:
        dataTypeString = 'float32';
        break;
      case TensorSerializerDataType.INT32:
        dataTypeString = 'int32';
        break;
      case TensorSerializerDataType.BOOL:
        dataTypeString = 'bool';
        break;
      case TensorSerializerDataType.UINT8:
        dataTypeString = 'uint8';
        break;
      default:
        throw new Error('Unsupported DataType');
    }
    const tensor = CPUTensor.fromArray(data, dims, dataTypeString);
    return { name, tensor };
  }

  private parseString(
    buf: ArrayBuffer,
    byteOffset: number,
    byteLength: number
  ): string {
    const view = new Uint8Array(buf, byteOffset, byteLength);
    // TODO: support UTF-8
    return String.fromCharCode(...Array.from(view));
  }

  private parseTensorBody(
    buf: ArrayBuffer,
    compressionAlgorithm: number,
    bodyByteOffset: number,
    bodyCompressedLength: number,
    dataType: number,
    numel: number
  ): TypedArrayTypes {
    switch (compressionAlgorithm) {
      case 0:
        return decodeTensorRaw(
          buf,
          bodyByteOffset,
          bodyCompressedLength,
          dataType,
          numel
        );
      default:
        throw new Error('Unexpected compression algorithm');
    }
  }
}

export class TensorSerializer {
  serialize(
    tensors: Map<string, CPUTensor> | Record<string, CPUTensor> | CPUTensor
  ): Uint8Array {
    let map: Map<string, CPUTensor>;
    if (tensors instanceof Map) {
      map = tensors;
    } else if (tensors instanceof CPUTensor) {
      map = new Map();
      map.set('default', tensors);
    } else {
      map = new Map();
      for (const [k, v] of Object.entries(tensors)) {
        map.set(k, v);
      }
    }
    return this.serializeCore(map);
  }

  async toHTTP(
    tensors: Map<string, CPUTensor> | Record<string, CPUTensor> | CPUTensor,
    path: string
  ): Promise<void> {
    const buf = this.serialize(tensors);
    const blob = new Blob([buf]);
    const f = await fetch(path, { method: 'POST', body: blob });
    if (!f.ok) {
      throw new Error('Server response to save is not ok');
    }
  }

  async toLocalStorage(
    tensors: Map<string, CPUTensor> | Record<string, CPUTensor> | CPUTensor,
    path: string
  ): Promise<void> {
    if (!path.startsWith(localStoragePrefix)) {
      throw new Error(
        `toLocalStorage: path must be prefix '${localStoragePrefix}'`
      );
    }
    const buf = this.serialize(tensors);
    // to support future splitting, save to path/0
    const key = path.substring(localStoragePrefix.length) + '/0';
    localStorage.setItem(key, uint8ArrayToBase64(buf));
  }

  /**
   * Serialize to download file.
   * @param tensors
   * @param fileName
   */
  async toFile(
    tensors: Map<string, CPUTensor> | Record<string, CPUTensor> | CPUTensor,
    fileName = 'tensors.bin'
  ): Promise<void> {
    const buf = this.serialize(tensors);
    const blob = new Blob([buf], { type: 'application/octet-stream' });
    const a = document.createElement('a');
    a.href = URL.createObjectURL(blob);
    a.download = fileName;

    a.style.display = 'none';
    document.body.appendChild(a);
    a.click();
    document.body.removeChild(a);
  }

  private serializeCore(tensors: Map<string, CPUTensor>): Uint8Array {
    let totalLength = 4;
    const tensorEntries: [string, CPUTensor, number, number][] = [];
    for (const [name, tensor] of tensors.entries()) {
      const chunkSize = this.calculateTensorChunkSize(name, tensor);
      tensorEntries.push([name, tensor, totalLength, chunkSize]);
      totalLength += chunkSize;
    }
    const closeChunkOffset = totalLength;
    totalLength += this.calculateCloseChunkSize();
    const dst = new Uint8Array(totalLength);
    const view = new DataView(dst.buffer);
    view.setUint32(0, signatureFile, true);
    for (const [name, tensor, offset, chunkSize] of tensorEntries) {
      this.makeTensorChunk(dst.buffer, offset, chunkSize, name, tensor);
    }
    this.makeCloseChunk(dst.buffer, closeChunkOffset);
    return dst;
  }

  private calculateTensorChunkSize(name: string, tensor: CPUTensor): number {
    // 固定長
    // signature: 4
    // chunk size: 4
    // compressionAlgorithm: 1
    // bodyCompressedLength: 4
    // dataType: 1
    // ndim: 1
    // nameLength: 4
    // extraLength: 4
    // extra: 0
    // 可変長
    // dims: 4 * tensor.ndim
    // name: name.length
    // data: tensor.data.byteLength
    if (!tensor.buffer) {
      throw new Error('buffer for tensor is null');
    }
    return 23 + name.length + tensor.ndim * 4 + tensor.buffer.data.byteLength;
  }

  private makeTensorChunk(
    dst: ArrayBuffer,
    dstOffset: number,
    chunkSize: number,
    name: string,
    tensor: CPUTensor
  ): void {
    const buffer = tensor.buffer;
    if (!buffer) {
      throw new Error('buffer for tensor is null');
    }
    const view = new DataView(dst, dstOffset);
    let ofs = 0;
    view.setUint32(ofs, signatureTensor, true);
    ofs += 4;
    view.setUint32(ofs, chunkSize - 8, true);
    ofs += 4;
    view.setUint8(ofs, 0); // compression algorithm
    ofs += 1;
    view.setUint32(ofs, buffer.data.byteLength, true); // compressed data length
    ofs += 4;
    let dataType: number;
    switch (tensor.dtype) {
      case 'float32':
        dataType = TensorSerializerDataType.FLOAT;
        break;
      case 'int32':
        dataType = TensorSerializerDataType.INT32;
        break;
      case 'uint8':
        dataType = TensorSerializerDataType.UINT8;
        break;
      case 'bool':
        dataType = TensorSerializerDataType.BOOL;
        break;
      default:
        throw new Error('Unsupported dtype');
    }
    view.setUint8(ofs, dataType);
    ofs += 1;
    view.setUint8(ofs, tensor.ndim);
    ofs += 1;
    for (const len of tensor.shape) {
      view.setUint32(ofs, len, true);
      ofs += 4;
    }
    const nameLength = name.length;
    view.setUint32(ofs, nameLength, true);
    ofs += 4;
    for (let i = 0; i < nameLength; i++) {
      view.setUint8(ofs, name.charCodeAt(i));
      ofs += 1;
    }
    view.setUint32(ofs, 0, true); //extraLength
    ofs += 4;
    const dstUint8 = new Uint8Array(dst, dstOffset + ofs);
    const srcUint8 = new Uint8Array(
      buffer.data.buffer,
      buffer.data.byteOffset,
      buffer.data.byteLength
    );
    dstUint8.set(srcUint8);
  }

  private calculateCloseChunkSize(): number {
    return 8;
  }

  private makeCloseChunk(dst: ArrayBuffer, dstOffset: number): void {
    const view = new DataView(dst, dstOffset);
    view.setUint32(0, signautreClose, true);
    view.setUint32(4, 0, true);
  }
}
